### PR TITLE
Tweak test APIs

### PR DESCRIPTION
**Pull request by jwoudenberg-nri**
I think this is good to go! Would love your thoughts. Motivations for this change are below, but concretely this PR has the following benefits:

- No more separate `Test.task` function. `Test.test` does what it did before (no changes in code necessary), but also supports tasks now.
- Clearer type errors when you attempt to run a Task in a test body with a custom error type.
- No more 'ambiguous type' errors when using `Expect.check` or `Expect.andCheck`
- No longer need to pass expectations through `Expect.check`.
- `Expect.andCheck` can be called without piping through `Expect.succeeds` first.
- `Expect.Task` functions in `Expect` so we have one fewer module to import
- No longer need to use `unsafePerformIO` in tests to create test handlers

---

I'd like to try tweak the test APIs a little bit. I think most of the functions are good and useful, but there's a couple of edges I think we could improve. Based on pairing with Michael today.

`Expect.task` tests run in the `Task Expect.Task.Failure a` monad. It's intended doing a mix of running tasks and making assertions. Both require conversion functions:

- Tasks typically have a different error type, so we need to pass them through `Expect.Task.succeeds` to plug them into the test.
- Expectations have a different type entirely, and need to be passed through `Expect.Task.check` to plug them into the test.

It would be nice if at least in one of these cases we wouldn't need to go through conversion functions. I think it's hard to implement this for the first case, because we'd have to settle on a single error type for `Expect.task` to be compatible with. Making it so expectations can be returned without conversion is easier and more broadly useful, and has as another bonus that it's compatible with regular `Test.test` which also expects an `Expectation` as a return value. This would mean we can drop `Test.task`, `Expect.Task` and add those couple of functions to `Expect`, saving an import.

Tasks would still need to be converted using `succeeds`, but there's a benefit here too. If we forget to pass the task through `succeeds` we now get a better type error, something like 'expected an Expectation but got a Task', rather than 'expected a Task Failure a but got a Task Error a`.

Still another example: We'd like tests to be able to create their own handles. This currently requires us to call `unsafePerformIO`. If our test body isn't a `Task` but some sort of `Expectation` type, we can provide a helper for handler construction `IO a -> Expectation' a` that is only useful in the context of tests.

In summary, I propose we keep the complete API (rejigging only some of the return types), but drop:
- Expect.Task.check (andCheck we should keep)
- Expect.Task module (move stuff into Expect`)

### CODE REVIEW

**Comment by stoeffel** [comment]
Cool I actually had Expectation be a Monad in a very early iteration, but for some reason dropped it again. It makes sense for tasks though!

**Comment by jwoudenberg-nri** [comment]
> Cool I actually had Expectation be a Monad in a very early iteration, but for some reason dropped it again.

I remember! I think back then we didn't support interleaving task evaluations and expectations which I think is why you designed `Test.task`. Now we can pick the best of both worlds!

**Comment by jwoudenberg-nri** [comment]
Added a couple of additional commits so we're tracing expectations too. This will make them show up in `log-explorer`.

File: nri-prelude/src/Expect.hs [hunk 1]
```diff
@@ -499,8 +510,107 @@ assert pred funcName actual expected =
...
+                  Internal.failAssertion
+                  "Expect.fails"
+                  (Debug.toString msg)
+      )
+    |> Internal.Expectation
+
+-- | This can be used to create custom test functions that contain some setup
+-- and teardown logic, for example to make tests run in a database transaction
+-- that gets rolled back afterwards.
+around ::
```
Comment (stoeffel) [inline nri-prelude/src/Expect.hs:605]: Would be nice to have an example in the docs for this.
Is it correct that it could be possible to forget to run the expectation this way?
What was the intention behind arround and not something like
```
before someTask <|
  after someOtherTask <|
  test "foo" <| \() ->
or
arround before after <| test "foo"
```

Comment (jwoudenberg-nri) [reply nri-prelude/src/Expect.hs:605]: The direct reason for adding this was making it possible to run tests in a test transaction. We need to be sure to clean up after the test completes, failure or not, and "around" patterns are I think the best way to ensure it happens (fits other haskell patterns too, like `bracket`).

Comment (jwoudenberg-nri) [reply nri-prelude/src/Expect.hs:605]: Also, this is personal preference, but I'm not a big fan of the before/after/around hooks of rspec or similar test frameworks, that encourage splitting up the test logic over a number of hooks, do things like mock data creation for a set of tests in a hook, etc.

I think making the `around` work on the `Expectation` rather than the `Test` makes it a good tool for creating custom versions of `Test.test` (for example: the ones we have running inside of a transaction), while not encouraging test body splitting.

Comment (stoeffel) [reply nri-prelude/src/Expect.hs:605]: got it! makes sense. definitely great to have this!

Comment (jwoudenberg-nri) [reply nri-prelude/src/Expect.hs:605]: > Is it correct that it could be possible to forget to run the expectation this way?

Yes, that's a good point. I hope that because it's such a rarely used function (for tooling creation rather than individual tests) the rick is limited though.

Comment (jwoudenberg-nri) [reply nri-prelude/src/Expect.hs:621]: Pushed an example!
